File: Phoenix-Framework/Server/Web/static/js/listeners.js
let original_modal_content = document.getElementById("create-modal-body").innerHTML;
let edit_listener_id = null;

function deleteListener(id) {
    fetch("/listeners/" + id + "/remove?json=true", {
        method: "DELETE"
    }).then(response => response.json())
        .then(data => {
            // show notification
            showNotification(data.message, data.status);
            // check if success
            if (data.status === "success") {
                // sleep 1 second
                setTimeout(function () {
                    // reload page
                    location.reload();
                }, 1000);
            }
        });
}

function startListener(id) {
    fetch("/listeners/" + id + "/start?json=true", {
        method: "POST"
    }).then(response => response.json())
        .then(data => {
            // show notification
            showNotification(data.message, data.status);
            // check if success
            if (data.status === "success") {
                // sleep 1 second
                setTimeout(function () {
                    // reload page
                    location.reload();
                }, 1000);
            }
        });
}

function restartListener(id) {
    showNotification(`Restarting listener ${id}.`, "info");
    // set listener activity status to false
    document.getElementById("active-" + id).innerHTML = '<i class="material-icons" style="color: red; margin-top: 4px;">circle</i>'
    fetch("/listeners/" + id + "/restart?json=true", {
        method: "POST"
    }).then(response => response.json())
        .then(data => {
            // show notification
            showNotification(data.message, data.status);
            // check if success
            if (data.status === "success") {
                // set listener activity status to true
                document.getElementById("active-" + id).innerHTML = '<i class="material-icons" style="color: green; margin-top: 4px;">circle</i>';
            }
        });
}

function stopListener(id) {
    fetch("/listeners/" + id + "/stop?json=true", {
        method: "POST"
    }).then(response => response.json())
        .then(data => {
            // show notification
            showNotification(data.message, data.status);
            // check if success
            if (data.status === "success") {
                // sleep 1 second
                setTimeout(function () {
                    // reload page
                    location.reload();
                }, 1000);
            }
        });
}

function resetModal() {
    // reset modal content
    document.getElementById("create-modal-body").innerHTML = original_modal_content;
    changeCreateType();
}
function changeCreateType() {
    // get type from select
    const type = document.getElementById("type").value;
    // get corresponding form
    const form = document.getElementById(type + "-form");
    // change content of modal
    document.getElementById("create-form-content").innerHTML = form.innerHTML 
}

function createEdit(id) {
    if (id == edit_listener_id) {
        // open modal
        $("#edit-modal").modal("show");
        return;
    }
    let listener = listeners[id];
    // get form by type
    const form = document.getElementById(listener.type + "-form");

    // set modal body
    // replace all create with form
    document.getElementById("edit-form").innerHTML = form.innerHTML.replace(/create/g, "edit");

    // set values
    document.getElementById("type-edit").remove();
    document.getElementById("id-edit").value = listener.id;
    document.getElementById("name-edit").value = listener.name;
    document.getElementById(listener.address + "-address-edit").selected = true;
    document.getElementById("port-edit").value = listener.port;
    document.getElementById("ssl-edit").checked = listener.ssl;
    document.getElementById("enabled-edit").checked = listener.enabled;
    document.getElementById("limit-edit").value = listener.limit;

    // set options
    for (let option_name in listener.options) {
        if (Object.prototype.hasOwnProperty.call(listener.options, option_name)) {
            let option = listener.options[option_name];
            let element = document.getElementById(option_name.toLowerCase() + "-edit");
            if (element.type === "checkbox") {
                element.checked = option;
            }
            element.value = option;
        }
    }

    // open modal
    $("#edit-modal").modal("show");
    console.log("lol");
}

File: Phoenix-Framework/tests/listeners.py
import unittest
import requests as r

# Test the Listener Endpoints of the API

json_param = '?json=true'

class ListenerTest(unittest.TestCase):
    url = 'http://localhost:8080/listeners/'
    session = r.Session()
    listener_id = "1"

    def test_get_listeners_1(self):
        print("Getting Listeners")
        response = self.session.get(self.url + json_param)
        self.assertEqual(response.status_code, 200)

    def test_create_listener_2(self):
        print("Creating Listener")
        data = {
            'name': 'test',
            'type': "http/reverse",
            'address': '0.0.0.0',
            'port': 9999,
            'ssl': True,
            'enabled': True,
            'limit': 0,
        }
        response = self.session.post(self.url + "/add" + json_param, data=data)
        self.assertEqual(response.status_code, 201)
        self.listener_id = str(response.json()['id'])


    def test_start_listener_3(self):
        print("Starting Listener")
        response = self.session.post(self.url + self.listener_id + "/start" + json_param, data={'id': 1})
        self.assertEqual(response.status_code, 200)
    
    def test_restart_listener_4(self):
        print("Restarting Listener")
        response = self.session.post(self.url + self.listener_id + "/restart" + json_param, data={'id': 1})
        self.assertEqual(response.status_code, 200)
    
    def test_stop_listener_5(self):
        print("Stopping Listener")
        response = self.session.post(self.url + self.listener_id + "/stop" + json_param, data={'id': 1})
        self.assertEqual(response.status_code, 200)
    
    
    def test_remove_listener_6(self):
        print("Removing Listener")
        response = self.session.delete(self.url + self.listener_id + "/remove" + json_param, data={'id': 1})
        self.assertEqual(response.status_code, 200)

if __name__ == '__main__':
    unittest.main()